src/librbd/plugin/Types.h

```diff
@@ -5,26 +5,36 @@
 #define CEPH_LIBRBD_PLUGIN_TYPES_H
 
 #include "include/common_fwd.h"
+#include "include/Context.h"
 #include "common/PluginRegistry.h"
 
-struct Context;
-
 namespace librbd {
 namespace plugin {
 
 template <typename> struct Api;
 
 struct HookPoints {
-  // TODO later commits will add support for exclusive-lock hook points
+  virtual ~HookPoints() {
+  }
+  virtual void acquire_exclusive_lock(Context* on_finish) = 0;
+  virtual void release_exclusive_lock(Context* on_finish) = 0;
+  virtual void discard(Context* on_finish) {
+    on_finish->complete(0);
+  }
 };
 
+typedef std::list<HookPoints*> PluginHookPoints;
+
 template <typename ImageCtxT>
 struct Interface : public ceph::Plugin {
   Interface(CephContext* cct) : Plugin(cct) {
   }
 
+  virtual ~Interface() {
+  }
+
   virtual void init(ImageCtxT* image_ctx, Api<ImageCtxT>& api,
-                    HookPoints* hook_points, Context* on_finish) = 0;
+                    PluginHookPoints& hook_points_list, Context* on_finish) = 0;
```

Comment:
@dillaman Do you think that I can change this interface?

Reply:
You can change any internal interface if needed. What's the plan?

Reply:
I updated the type of m_plugin_hook_points in librbd::PluginRegister to the following type:
typedef std::list<std::unique_ptr<HookPoints>> PluginHookPoints;

And pass it to the above init function, so that the hooks can be added directly to the list.